server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java

```diff
@@ -86,34 +89,32 @@ public class GarbageCollectWriteAheadLogs {
     this.fs = fs;
     this.useTrash = useTrash;
     this.liveServers = liveServers;
-    this.walMarker = new WalStateManager(context);
-    this.store = () -> Iterators.concat(
-        TabletStateStore.getStoreForLevel(DataLevel.ROOT, context).iterator(),
-        TabletStateStore.getStoreForLevel(DataLevel.METADATA, context).iterator(),
-        TabletStateStore.getStoreForLevel(DataLevel.USER, context).iterator());
+    this.walMarker = createWalStateManager(context);
+    this.hasCollected = new AtomicBoolean(false);
   }
 
-  /**
-   * Creates a new GC WAL object. Meant for testing -- allows mocked objects.
-   *
-   * @param context the collection server's context
-   * @param fs volume manager to use
-   * @param useTrash true to move files to trash rather than delete them
-   * @param liveTServerSet a started LiveTServerSet instance
-   */
   @VisibleForTesting
-  GarbageCollectWriteAheadLogs(ServerContext context, VolumeManager fs, boolean useTrash,
-      LiveTServerSet liveTServerSet, WalStateManager walMarker,
-      Iterable<TabletLocationState> store) {
-    this.context = context;
-    this.fs = fs;
-    this.useTrash = useTrash;
-    this.liveServers = liveTServerSet;
-    this.walMarker = walMarker;
-    this.store = store;
+  WalStateManager createWalStateManager(ServerContext context) {
+    return new WalStateManager(context);
+  }
+
+  @VisibleForTesting
+  Stream<TabletLocationState> createStore() {
+    Stream<TabletLocationState> rootStream =
+        TabletStateStore.getStoreForLevel(DataLevel.ROOT, context).stream();
+    Stream<TabletLocationState> metadataStream =
+        TabletStateStore.getStoreForLevel(DataLevel.METADATA, context).stream();
+    Stream<TabletLocationState> userStream =
+        TabletStateStore.getStoreForLevel(DataLevel.USER, context).stream();
+    return Streams.concat(rootStream, metadataStream, userStream).onClose(() -> {
+      rootStream.close();
+      metadataStream.close();
+      userStream.close();
```

Comment:
If these throw a RTE, you probably want to continue trying to close the others. I don't know a clean way to express that... maybe something like:

```suggestion
      try {
        rootStream.close();
      } finally {
        try {
          metadataStream.close();
        } finally {
          userStream.close();
        }
      }
```